packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -424,13 +424,15 @@ export default class WorkspacesForeachCommand extends BaseCommand {
           return;
         }
 
-        const exitCodes: Array<number> = await Promise.all(commandPromises);
-        const errorCode = exitCodes.find(code => code !== 0);
+        const results: Array<{ workspace: Workspace, exitCode: number }> = await Promise.all(commandPromises);
+        results.forEach(({workspace, exitCode}) => {
+          if (exitCode !== 0) {
+            report.reportError(MessageName.UNNAMED, `The command failed in workspace ${structUtils.prettyLocator(configuration, workspace.anchoredLocator)} with exit code ${exitCode}`);
+          }
+        });
 
-        // The order in which the exit codes will be processed is fairly
-        // opaque, so better just return a generic "1" for determinism.
-        if (finalExitCode === null)
-          finalExitCode = typeof errorCode !== `undefined` ? 1 : finalExitCode;
```

Comment:
we do not need to set the `finalExitCode` bc the resulting error code will be set to 1 by `report.exitCode()` since errors have been reported.

https://github.com/yarnpkg/berry/blob/80052a0736bb7c2d3ed9b7c7a121a2983a41bee0/packages/yarnpkg-core/sources/StreamReport.ts#L267-L269